hivemind/utils/mpfuture.py

```diff
@@ -60,49 +63,36 @@ class MPFuture(base.Future, Generic[ResultType]):
+    _status_requests: Optional[Dict[UID, Tuple[MPFuture, threading.Event]]] = None  # futures to be updated by origin
     _active_pid: Optional[PID] = None  # pid of currently active process; used to handle forks natively
 
-    def __init__(self, use_lock: bool = True, loop: Optional[asyncio.BaseEventLoop] = None):
-        self._origin_pid, self._uid = os.getpid(), uuid.uuid4().int
-        self._shared_state_code = torch.empty([], dtype=torch.uint8).share_memory_()
-        self._state_cache: Dict[State, State] = {}
-        # mapping from global to cached local future used that makes updates immediately
-        # available on setter side; dictionary-based cache works because future can visit any state at most once
+    SOFT_UPDATE_TIMEOUT = 0.1  # seconds spent awaiting status update before warning is printed
+    HARD_UPDATE_TIMEOUT = 10.0  # seconds spent awaiting status update before future is automatically cancelled
 
-        base.Future.__init__(self)  # parent init is deferred because it uses self._shared_state_code
+    def __init__(self, synchronize: bool = True, use_lock: bool = True, loop: Optional[asyncio.BaseEventLoop] = None):
+        base.Future.__init__(self)
+        self.synchronize = synchronize
+        self._origin_pid, self._uid = os.getpid(), uuid.uuid4().int
         self._state, self._result, self._exception = base.PENDING, None, None
         self._use_lock = use_lock
 
-        if self._origin_pid != MPFuture._active_pid:
-            with MPFuture._initialization_lock:
-                if self._origin_pid != MPFuture._active_pid:
-                    # note: the second if is intentional, see https://en.wikipedia.org/wiki/Double-checked_locking
-                    self._initialize_mpfuture_backend()
+        self._initialize_backend_if_necessary()
         assert self._uid not in MPFuture._active_futures
-        MPFuture._active_futures[self._uid] = self
-        self._sender_pipe = MPFuture._global_sender_pipe
+        MPFuture._active_futures[self._uid] = ref(self)
+        self._sender_pipe = MPFuture._process_wide_pipe
 
         try:
             self._loop = loop or asyncio.get_event_loop()
             self._aio_event = asyncio.Event()
         except RuntimeError:
             self._loop, self._aio_event = None, None
 
-    @property
-    def _state(self) -> State:
-        shared_state = ALL_STATES[self._shared_state_code.item()]
-        return self._state_cache.get(shared_state, shared_state)
-
-    @_state.setter
-    def _state(self, new_state: State):
-        self._shared_state_code[...] = ALL_STATES.index(new_state)
-        if self._state in TERMINAL_STATES and self._loop is not None and not self._aio_event.is_set():
-            self._set_event_threadsafe()
-
-    def _set_event_threadsafe(self):
+    def _set_event_if_necessary(self):
+        if self._aio_event is None or self._aio_event.is_set():
```

Comment:
`self._aio_event.is_set()` is not thread-safe.

Reply:
decided to keep is_set for performance reasons

Reply:
[It's thread-safe in the current implementation, and we've agreed that it's hard to imagine that it will stop to be thread-safe.]